transfer/rsync/client.go

```diff
@@ -0,0 +1,431 @@
+	labels map[string]string,
+	ownerRefs []metav1.OwnerReference,
+	password string, podOptions transfer.PodOptions) (transfer.Client, error) {
+	tc := &client{
+		username:        "root",
+		password:        password,
+		pvcList:         pvcList,
+		transportClient: t,
+		endpoint:        e,
+		nameSuffix:      nameSuffix,
+		labels:          labels,
+		ownerRefs:       ownerRefs,
+		options:         podOptions,
+	}
+
+	var namespace string
+	namespaces := pvcList.Namespaces()
+	if len(namespaces) > 0 {
+		namespace = pvcList.Namespaces()[0]
+	}
+
+	for _, ns := range namespaces {
+		if ns != namespace {
+			return nil, fmt.Errorf("PVC list provided has pvcs in different namespaces which is not supported")
+		}
+	}
+
+	if namespace == "" {
+		return nil, fmt.Errorf("ether PVC list is empty or namespace is not specified")
+	}
+	tc.namespace = namespace
+
+	tc.nameSuffix = transfer.NamespaceHashForNames(pvcList)[namespace][:10]
+	reconcilers := []reconcileFunc{
+		tc.reconcileServiceAccount,
+		tc.reconcileRole,
+		tc.reconcileRoleBinding,
+		tc.reconcilePod,
+	}
+
+	for _, reconcile := range reconcilers {
+		err := reconcile(ctx, c, tc.namespace)
+		if err != nil {
+			tc.logger.Error(err, "error reconciling rsyncServer")
+			return nil, err
+		}
+	}
+
+	return tc, nil
+}
+
+// TODO: add retries
+func (tc *client) reconcilePod(ctx context.Context, c ctrlclient.Client, ns string) error {
+	var errs []error
+	rsyncOptions, err := GetRsyncCommandDefaultFlags()
+	if err != nil {
+		tc.logger.Error(err, "unable to get default flags for rsync command")
+		return err
+	}
+	for _, pvc := range tc.pvcList.InNamespace(ns).PVCs() {
+		// create Rsync command for PVC
+		rsyncContainerCommand := tc.getCommand(rsyncOptions, pvc)
+		// create rsync container
+		containers := []corev1.Container{
+			{
+				Name:    RsyncContainer,
+				Image:   rsyncImage,
+				Command: rsyncContainerCommand,
+				Env: []corev1.EnvVar{
+					{
+						Name:  "RSYNC_PASSWORD",
+						Value: tc.password,
+					},
+				},
+
+				VolumeMounts: []corev1.VolumeMount{
+					{
+						Name:      "mnt",
+						MountPath: fmt.Sprintf("/mnt/%s/%s", pvc.Claim().Namespace, pvc.LabelSafeName()),
+					},
+					{
+						Name:      "rsync-communication",
+						MountPath: "/usr/share/rsync",
+					},
+				},
+			},
+		}
+		// attach transport containers
+		err := customizeTransportClientContainers(tc.Transport())
+		if err != nil {
+			tc.logger.Error(err, "unable to customize Transport client containers for rsync client pod")
+			return err
+		}
+		containers = append(containers, tc.Transport().Containers()...)
+
+		volumes := []corev1.Volume{
+			{
+				Name: "mnt",
+				VolumeSource: corev1.VolumeSource{
+					PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+						ClaimName: pvc.Claim().Name,
+					},
+				},
+			},
+			{
+				Name: "rsync-communication",
+				VolumeSource: corev1.VolumeSource{
+					EmptyDir: &corev1.EmptyDirVolumeSource{Medium: corev1.StorageMediumDefault},
```

Comment:
Assuming this is just a socket:
```suggestion
					EmptyDir: &corev1.EmptyDirVolumeSource{Medium: corev1.StorageMediumMemory},
```